File: 01_code.py
#!/usr/bin/env python3

def divider (value , divisor):
    result = value // divisor
    return result;

def substractor (value, substract):
    result = value - substract
    if result > 0:
        return result
    else:
        return 0

def summarizer (values):
    required_fuel = 0
    for v in values:
        divided = divider(v, 3)
        fuel = substractor(divided, 2)
        required_fuel += fuel
    return required_fuel

def read_integers(filename):
    with open(filename) as f:
        return list(map(int, f))

def main():
    vals = read_integers("01_data.txt")
    total_fuel = summarizer(vals)
    print("Total fuel required: ", total_fuel)
    return

if __name__=="__main__":
    main()

File: 01_code.lua
#!/usr/bin/env lua

------------------------------------------------------------------------------------------

-- This class is aimed to be used for reading of numbers from a file.
local Reader = {}
Reader.new = function(self, object)
  object = object or {}
  setmetatable(object, self)
  self.__index = self
  return object
end

---@diagnostic disable-next-line: unused-local
Reader.__call = function(self, file_name)
  if(type(file_name) ~= "string") then
    print("Please provide file containing numbers as a string!")
    return
  end

  local file = assert(
    io.open(file_name, "r"),
    "Failed to open file " .. file_name .. " for reading!")
  local lines = file:lines()
  local numbers = {}
  for line in lines do
    local number = tonumber(line)
    if(number) then
      table.insert(numbers, number)
    else
      print("Error converting line: " .. line)
    end
  end
  file:close()
  return numbers
end

------------------------------------------------------------------------------------------

local input    = "./01_data.txt"
local reader  = Reader:new()
local numbers = reader(input)
local desired = 2020

while(#numbers > 2) do
  print("Remaining ... " .. #numbers)
  local base = table.remove(numbers, 1) -- base contains the first table element (that will be dropped in the ext iteration)
  local size = #numbers
  local key = 1
  while(key + 2 < size) do
    local idx = key + 1
    while(idx + 1 < size) do
      local sum = base + numbers[key] + numbers[idx]
      if(desired == sum) then
        local mul = base * numbers[key] * numbers[idx]
        print(sum .. " = " .. base .. " + " .. numbers[key] .. " + " .. numbers[idx])
        print(mul .. " = " .. base .. " * " .. numbers[key] .. " + " .. numbers[idx])
        return
      end
      idx = idx + 1
    end
    key = key + 1
  end
end
